docs/src/gradient_accumulation.md

```diff
@@ -0,0 +1,132 @@
+# Gradient Accumulation
+
+Consider some function
+$$f(x) = g(x) + h(x)$$.
+If we would like the derivative of $f$ with respect to $x$ we must compute it for each part and then sum them, i.e. 
+$$\frac{\partial f}{\partial x} = \frac{\partial g}{\partial x} + \frac{\partial h}{\partial x}$$.
+In general, we must accumulate (sum) gradients from each sub-part of a program where a variable is used.
+
+
+Consider for example:
+```julia
+function sum_first_and_second(X::Array{Float64})
+    a = X[1]
+    b = X[2]
+    y = a + b
+    return y
+end
+```
+The AD software must transform that into something which repeatedly sums up the gradient of each part:
+`X̄ = ā + b̄`.
+
+This requires that all differential types `D` must implement `+`: `+(::D, ::D)::D`.
+
+We can note that in this particular case `ā` and `b̄` will both be arrays.
+This operation (`X̄ = ā + b̄`) will allocate one array to hold `ā`, another one to hold `b̄`, and a third one to hold `ā + b̄`.
+This is three allocations.
+Allocations are not free, they increase the time the program takes to run by a nontrivial amount, even with a good allocator and a good garbage collector.
+
+### Maybe-mutating accumulation (`add!!`)
+We can note that in the above that neither `ā` nor `b̄` are ever used again after accumulating to get `X̄`.
+Furthermore, `Array`s are mutable.
+That means we could over-write either `ā` or `b̄` and use the result as `X̄`:
+```julia
+ā .+= b̄
+X̄ = ā
+```
+This cuts our allocations down to 2, just `ā` and `b̄`.
+
+However, we have a bit of a problem that not all types are mutable, so this pattern is hard to apply in general.
+To deal with that ChainRulesCore provides [`add!!`](@ref).
+Per the [BangBang.jl](https://github.com/JuliaFolds/BangBang.jl) convention, this is a maybe mutating addition.
+It may mutate its first argument (it it is mutable), but it will definitely return the correct result.
+We would write using that as `X̄ = add!!(ā, b̄)`: which would in this case give us just 2 allocations.
+AD systems can generate `add!!` instead of `+` when accumulating gradient to take advantage of this.
+
+### Inplaceable Thunks (`InplaceableThunks`) avoid allocating values in the first place.
+We got down to two allocations from using [`add!!`](@ref), but can we do better?
+We can think of having a differential type which acts on a partially accumulated result, to mutate it to contain its current value plus the partial derivative being accumulated.
+Rather than having an actual computed value, we can just have a thing that will act on a value to perform the addition.
+Let's illustrate it with our example.
+
+`b̄` is the partial for `X[2]` and its value can be computed by:
+```julia
+b̄ = zeros(size(X))
+b̄[2] = ȳ  # the scalar sensitivity of the `mysum` output
+```
```

Comment:
```suggestion

```julia
b̄ = zeros(size(X))
b̄[2] = ȳ  # the scalar sensitivity of the `mysum` output
```

```